### PR TITLE
Improve self tests

### PR DESCRIPTION
**Pull request by Laar**
After realizing that updating the results in the self tests that fail in #123 was very complicated I decided to refactor these first. 

There were two problems with these self tests:
 1. The error (and in general the errror in the API classes) was computed using a too low quadrature rule
 2. The computations used varied both the mesh size and order of the polynomials used in the basis functions.

The first is a problem as the error is usually underestimated when using a quadrature rule that would be 'just accurate enough'. Using a quadrature rule of slightly higher order gives a result that is a far closer value. Here we therefore switch to using 2 orders above the polynomial order of the basis functions.

The second is a problem as this prevents comparing the numerical error to see if the solutions are accurate enough. This is now replaced by varying only the mesh size but keeping the polynomial order fixed. As result one can easily see if the decay in the errors follows the expected power laws e.g. h^{p+1}. To simplify this approach two extra classes have been added:
 1. `TestMeshes` which provides easy access to a set of test meshes, which are both documented and don't need to be duplicated between tests
 2. `ConvergenceTest` to prevent duplicating the process of solving the same problem on multiple meshes and computing the convergence coefficients.

Some test cases were modified to speed up convergence or use different boundary condition (080LinearAdvection).

### CODE REVIEW

**Review by JensWehner — COMMENTED** [review]
Small improvements, not a full review

File: kernel/Output/VTKTimeDependentWriter.h [hunk 1]
```diff
@@ -93,6 +95,7 @@ class VTKTimeDependentWriter final {
     std::size_t timelevel_;
     double time_;
     std::size_t numberOfFilesWritten_;
+    std::size_t order_;
```
[the remark this answers is not in the document]
Comment (Laar) [reply kernel/Output/VTKTimeDependentWriter.h:98]: Actually not, quadrature has to do with integration. This is a combination of two things:
 - Possible curvature of the sides of the plotted elements. Are they straight lines (order=1), quadratic/bilinear curves (order=2), etc.
 - The interpolation of the solution, e.g. is it exported as linear function (order=1), is it plotted as a quadratic or bilinear function (order=2), etc.

So might change it to polynomialOrder.

File: tests/self/Basic/080AdvectionLinear_SelfTest.cpp [hunk 2]
```diff
@@ -191,58 +222,80 @@ class AdvectionLinear : public Base::HpgemAPILinear<DIM> {
...
-            // "comparison to old results");
-        }
-    }
+
+    // For recomputing the error tables
+    bool ignoreFailures = false;
+
+    AdvectionParameters dim1Params = {1, 0.1, 10};
+    // Minimum of 4 elements
+    ConvergenceTestSet dim1Meshes = {getUnitSegmentMeshes(2),
```
[the remark this answers is not in the document]
Comment (Laar) [reply tests/self/Basic/080AdvectionLinear_SelfTest.cpp:257]: I don't understand what you specifically mean.

They are just the mesh files included in the git repository. Each of them was generated using the preprocessor (just like the predecessors). But now their content is clear from the function name/documentation and they can be shared if needed.

File: kernel/Output/VTKTimeDependentWriter.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (Laar) [reply kernel/Output/VTKTimeDependentWriter.h]: Done

File: tests/self/Basic/080AdvectionLinear_SelfTest.cpp [hunk 3]
```diff
@@ -66,12 +76,33 @@ class AdvectionLinear : public Base::HpgemAPILinear<DIM> {
 
     /// Constructor. Assign all private variables.
     AdvectionLinear(const std::string fileName, const std::size_t p)
-        : Base::HpgemAPILinear<DIM>(1, p), fileName(fileName) {
+        : Base::HpgemAPILinear<DIM>(
+              1, p,
+              TimeIntegration::AllTimeIntegrators::Instance().getRule(5, 7), 0,
+              false, true),
```
[the remark this answers is not in the document]
Comment (Laar) [reply tests/self/Basic/080AdvectionLinear_SelfTest.cpp:82]: It definitely is, but that is also quite beyond this PR. What you see here is just inlining the default argument so that I can change the last one.

File: tests/self/Basic/080AdvectionLinear_SelfTest.cpp [hunk 4]
```diff
@@ -191,58 +222,80 @@ class AdvectionLinear : public Base::HpgemAPILinear<DIM> {
...
+    runAdvectionTestSet<2>(dim2P2Meshes, dim2P2Params, ignoreFailures);
+
+    // First order dim 3
+    ConvergenceTestSet dim3Meshes = {getUnitCubeCubeMeshes(2, 5),
+                                     {
+                                         5.55152172e-02,  //------
+                                         1.43980436e-02,  //  3.86
+                                         3.81712900e-03,  //  3.77
+                                     }};
+    AdvectionParameters dim3Params = {1, 0.02, 2};
```
[the remark this answers is not in the document]
Comment (Laar) [reply tests/self/Basic/080AdvectionLinear_SelfTest.cpp:296]: What/where specifically, because every file is now documented in what numerical problem it solves and what convergence rate to expect. The tables as you see them here are rather standard in FEM literature (though there with less precision).

I did add a line in ConvergenceTest to hint at its intended usage.

[the remark this answers is not in the document]
Comment (JensWehner) [reply tests/self/Basic/080AdvectionLinear_SelfTest.cpp:296]: thanks

**Review by JensWehner — APPROVED** [review]
Good work
